File: android/src/main/kotlin/app/capgo/auto/Auto.kt
package app.capgo.auto

import android.content.Context
import com.getcapacitor.JSObject
import java.lang.ref.WeakReference
import java.util.concurrent.CopyOnWriteArrayList
import org.json.JSONArray
import org.json.JSONObject

internal const val ROOT_TEMPLATE_STATE_KEY = "__capgo_auto_root_template"

data class AutoTemplateItem(
    val id: String,
    val title: String,
    val subtitle: String?,
    val payload: JSONObject?,
    val enabled: Boolean,
)

data class AutoTemplateSection(
    val header: String?,
    val items: List<AutoTemplateItem>,
)

data class AutoTemplate(
    val title: String,
    val sections: List<AutoTemplateSection>,
    val emptyText: String,
) {
    fun toJson(): JSObject {
        val sectionsJson = JSONArray()
        sections.forEach { section ->
            val sectionJson = JSONObject()
            section.header?.let { sectionJson.put("header", it) }

            val itemsJson = JSONArray()
            section.items.forEach { item ->
                val itemJson = JSONObject().apply {
                    put("id", item.id)
                    put("title", item.title)
                    item.subtitle?.let { put("subtitle", it) }
                    item.payload?.let { put("payload", it) }
                    put("enabled", item.enabled)
                }
                itemsJson.put(itemJson)
            }

            sectionJson.put("items", itemsJson)
            sectionsJson.put(sectionJson)
        }

        return JSObject().apply {
            put("title", title)
            put("sections", sectionsJson)
            put("emptyText", emptyText)
        }
    }

    companion object {
        val fallback = AutoTemplate(
            title = "Auto",
            sections = emptyList(),
            emptyText = "Open the app to configure Auto.",
        )

        fun fromJson(rawTemplate: JSONObject?): AutoTemplate? {
            if (rawTemplate == null) {
                return null
            }

            val title = rawTemplate.optString("title").takeIf { it.isNotBlank() } ?: return null
            return AutoTemplate(
                title = title,
                sections = parseSections(rawTemplate.optJSONArray("sections")),
                emptyText = rawTemplate.optString("emptyText").takeIf { it.isNotBlank() } ?: "No actions available.",
            )
        }

        private fun parseSections(rawSections: JSONArray?): List<AutoTemplateSection> {
            if (rawSections == null) {
                return emptyList()
            }

            return buildList {
                for (sectionIndex in 0 until rawSections.length()) {
                    val rawSection = rawSections.optJSONObject(sectionIndex) ?: continue
                    add(
                        AutoTemplateSection(
                            header = rawSection.optString("header").takeIf { it.isNotBlank() },
                            items = parseItems(rawSection.optJSONArray("items")),
                        ),
                    )
                }
            }
        }

        private fun parseItems(rawItems: JSONArray?): List<AutoTemplateItem> {
            if (rawItems == null) {
                return emptyList()
            }

            return buildList {
                for (itemIndex in 0 until rawItems.length()) {
                    val rawItem = rawItems.optJSONObject(itemIndex) ?: continue
                    val id = rawItem.optString("id")
                    val title = rawItem.optString("title")

                    if (id.isBlank() || title.isBlank()) {
                        continue
                    }

                    add(
                        AutoTemplateItem(
                            id = id,
                            title = title,
                            subtitle = rawItem.optString("subtitle").takeIf { it.isNotBlank() },
                            payload = rawItem.optJSONObject("payload"),
                            enabled = if (rawItem.has("enabled")) rawItem.optBoolean("enabled") else true,
                        ),
                    )
                }
            }
        }
    }
}

class Auto {
    fun getPluginVersion(): String {
        return "native"
    }
}

internal object AutoBridge {
    private var plugin: AutoPlugin? = null
    private var store: AutoStore? = null
    private val pendingEvents = mutableListOf<Pair<String, JSObject>>()
    private val screens = CopyOnWriteArrayList<WeakReference<AutoScreen>>()
    private val storeListener = AutoStore.Listener { key, value, transient ->
        if (key != ROOT_TEMPLATE_STATE_KEY && plugin != null) {
            emitStateChanged(key, value, transient)
        }
    }

    var template: AutoTemplate = AutoTemplate.fallback
        private set

    var connected: Boolean = false
        private set

    var lastMessage: JSObject? = null
        private set

    fun attach(plugin: AutoPlugin) {
        this.plugin = plugin
        emitConnection()

        pendingEvents.forEach { (name, data) ->
            plugin.emitEvent(name, data)
        }
        pendingEvents.clear()
    }

    fun detach(plugin: AutoPlugin) {
        if (this.plugin === plugin) {
            this.plugin = null
        }
    }

    fun configure(context: Context) {
        val nextStore = AutoStore.get(context)
        if (store !== nextStore) {
            store?.removeListener(storeListener)
            store = nextStore
            nextStore.addListener(storeListener)
        }

        AutoTemplate.fromJson(nextStore.load(ROOT_TEMPLATE_STATE_KEY))?.let {
            template = it
        }
    }

    fun registerScreen(screen: AutoScreen) {
        screens.add(WeakReference(screen))
    }

    fun setTemplate(template: AutoTemplate) {
        this.template = template
        store?.save(ROOT_TEMPLATE_STATE_KEY, template.toJson(), synchronous = true)
        refreshScreens()
    }

    fun setState(key: String, value: JSONObject) {
        requireStore().save(key, value)
    }

    fun getState(key: String): JSONObject? {
        return requireStore().load(key)
    }

    fun removeState(key: String) {
        requireStore().remove(key)
    }

    fun setTransientState(key: String, value: JSONObject) {
        requireStore().setTransient(key, value)
    }

    fun getTransientState(key: String): JSONObject? {
        return requireStore().getTransient(key)
    }

    fun setConnected(connected: Boolean) {
        this.connected = connected
        emitConnection()
    }

    fun receiveAction(item: AutoTemplateItem) {
        val data = JSObject().apply {
            put("id", item.id)
            put("title", item.title)
            put("platform", "android")
            item.payload?.let { put("payload", it) }
        }

        emit("carAction", data)
    }

    fun receiveMessage(type: String, payload: JSONObject?) {
        val data = JSObject().apply {
            put("type", type)
            put("platform", "android")
            payload?.let { put("payload", it) }
        }

        lastMessage = data
        emit("messageReceived", data)
    }

    private fun emitStateChanged(key: String, value: JSONObject?, transient: Boolean) {
        val data = JSObject().apply {
            put("key", key)
            put("platform", "android")
            put("transient", transient)
            value?.let { put("value", it) }
        }
        emit("stateChanged", data)
    }

    private fun emitConnection() {
        val data = JSObject().apply {
            put("connected", connected)
            put("platform", "android")
        }
        emit("connectionChanged", data)
    }

    private fun emit(eventName: String, data: JSObject) {
        val currentPlugin = plugin
        if (currentPlugin == null) {
            pendingEvents.add(eventName to data)
            return
        }

        currentPlugin.emitEvent(eventName, data)
    }

    private fun refreshScreens() {
        screens.removeAll { reference ->
            val screen = reference.get()
            if (screen == null) {
                true
            } else {
                screen.refresh()
                false
            }
        }
    }

    private fun requireStore(): AutoStore {
        return store ?: throw IllegalStateException("AutoBridge is not configured with an Android context")
    }
}

File: ios/Sources/AutoPlugin/Auto.swift
import Foundation

public struct AutoTemplateItem {
    let id: String
    let title: String
    let subtitle: String?
    let payload: [String: Any]?
    let enabled: Bool
}

public struct AutoTemplateSection {
    let header: String?
    let items: [AutoTemplateItem]
}

public struct AutoTemplate {
    let title: String
    let sections: [AutoTemplateSection]
    let emptyText: String

    static let fallback = AutoTemplate(
        title: "Auto",
        sections: [],
        emptyText: "Open the app to configure Auto."
    )

    func toDictionary() -> [String: Any] {
        [
            "title": title,
            "sections": sections.map { section in
                var sectionData: [String: Any] = [
                    "items": section.items.map { item in
                        var itemData: [String: Any] = [
                            "id": item.id,
                            "title": item.title,
                            "enabled": item.enabled
                        ]

                        if let subtitle = item.subtitle {
                            itemData["subtitle"] = subtitle
                        }
                        if let payload = item.payload {
                            itemData["payload"] = payload
                        }

                        return itemData
                    }
                ]

                if let header = section.header {
                    sectionData["header"] = header
                }

                return sectionData
            },
            "emptyText": emptyText
        ]
    }

    static func fromDictionary(_ rawTemplate: [String: Any]) -> AutoTemplate? {
        guard let title = rawTemplate["title"] as? String, !title.isEmpty else {
            return nil
        }

        return AutoTemplate(
            title: title,
            sections: parseSections(rawTemplate["sections"] as? [Any] ?? []),
            emptyText: rawTemplate["emptyText"] as? String ?? "No actions available."
        )
    }

    private static func parseSections(_ rawSections: [Any]) -> [AutoTemplateSection] {
        rawSections.compactMap { rawSection in
            guard let section = rawSection as? [String: Any] else {
                return nil
            }

            let rawItems = section["items"] as? [[String: Any]] ?? []
            let items = rawItems.compactMap(parseItem)

            return AutoTemplateSection(
                header: section["header"] as? String,
                items: items
            )
        }
    }

    private static func parseItem(_ rawItem: [String: Any]) -> AutoTemplateItem? {
        guard
            let id = rawItem["id"] as? String,
            let title = rawItem["title"] as? String,
            !id.isEmpty,
            !title.isEmpty
        else {
            return nil
        }

        return AutoTemplateItem(
            id: id,
            title: title,
            subtitle: rawItem["subtitle"] as? String,
            payload: rawItem["payload"] as? [String: Any],
            enabled: rawItem["enabled"] as? Bool ?? true
        )
    }
}

extension Notification.Name {
    static let capgoAutoTemplateChanged = Notification.Name("CapgoAutoTemplateChanged")
}

public final class AutoBridge {
    public static let shared = AutoBridge()

    private static let rootTemplateStateKey = "__capgo_auto_root_template"

    private weak var plugin: AutoPlugin?
    private var pendingEvents: [(String, [String: Any])] = []

    public private(set) var template = AutoTemplate.fallback
    public private(set) var connected = false
    public private(set) var lastMessage: [String: Any]?

    private init() {
        AutoStore.shared.addListener(self)
        loadStoredTemplate()
    }

    func attach(_ plugin: AutoPlugin) {
        self.plugin = plugin
        emit("connectionChanged", data: [
            "connected": connected,
            "platform": "ios"
        ])

        pendingEvents.forEach { name, data in
            plugin.emitEvent(name, data: data)
        }
        pendingEvents.removeAll()
    }

    func detach(_ plugin: AutoPlugin) {
        if self.plugin === plugin {
            self.plugin = nil
        }
    }

    func setTemplate(_ template: AutoTemplate) {
        self.template = template
        AutoStore.shared.save(Self.rootTemplateStateKey, template.toDictionary())
        NotificationCenter.default.post(name: .capgoAutoTemplateChanged, object: nil)
    }

    func reloadStoredTemplate() {
        loadStoredTemplate()
    }

    func setState(key: String, value: [String: Any]) {
        AutoStore.shared.save(key, value)
    }

    func getState(key: String) -> [String: Any]? {
        AutoStore.shared.load(key)
    }

    func removeState(key: String) {
        AutoStore.shared.remove(key)
    }

    func setTransientState(key: String, value: [String: Any]) {
        AutoStore.shared.setTransient(key, value)
    }

    func getTransientState(key: String) -> [String: Any]? {
        AutoStore.shared.getTransient(key)
    }

    func setConnected(_ connected: Bool) {
        self.connected = connected
        emit("connectionChanged", data: [
            "connected": connected,
            "platform": "ios"
        ])
    }

    func receiveAction(_ item: AutoTemplateItem) {
        var data: [String: Any] = [
            "id": item.id,
            "title": item.title,
            "platform": "ios"
        ]

        if let payload = item.payload {
            data["payload"] = payload
        }

        emit("carAction", data: data)
    }

    func receiveMessage(type: String, payload: [String: Any]?) {
        var data: [String: Any] = [
            "type": type,
            "platform": "ios"
        ]

        if let payload = payload {
            data["payload"] = payload
        }

        lastMessage = data
        emit("messageReceived", data: data)
    }

    private func emit(_ name: String, data: [String: Any]) {
        guard let plugin else {
            pendingEvents.append((name, data))
            return
        }

        plugin.emitEvent(name, data: data)
    }

    private func loadStoredTemplate() {
        guard
            let storedTemplate = AutoStore.shared.load(Self.rootTemplateStateKey),
            let template = AutoTemplate.fromDictionary(storedTemplate)
        else {
            return
        }

        self.template = template
    }
}

extension AutoBridge: AutoStoreListener {
    public func onAutoStoreUpdated(_ key: String, value: [String: Any]?, transient: Bool) {
        guard key != Self.rootTemplateStateKey else {
            return
        }

        var data: [String: Any] = [
            "key": key,
            "platform": "ios",
            "transient": transient
        ]

        if let value = value {
            data["value"] = value
        }

        emit("stateChanged", data: data)
    }
}

@objc public class Auto: NSObject {
    @objc public func getPluginVersion() -> String {
        return "native"
    }
}
